### PR TITLE
WIP: Parse endpoints using argparse

### PR DESCRIPTION
**Pull request by jayvdb**

### CODE REVIEW

**Comment by jayvdb** [comment]
Happy to re-add `find_endpoint` and mark it as deprecated somehow.
Also happy to add more tests to help guarantee this is working correctly for any edge cases.

**Comment by cmhulbert** [comment]
This is really coincidental; I've run into this issue also the last few days, and just last night (shortly before this issue), I opened a PR to fix this as well https://github.com/scijava/jgo/pull/74

I take a different approach though, just adding a patter to match window's drive letters and ignore, similar to how it already did for `http://`

File: jgo/jgo.py [hunk 1]
```diff
@@ -717,15 +719,18 @@ def run(parser, argv=sys.argv[1:], stdout=None, stderr=None):
...
+        raise HelpRequested(argv)
+
+    args, unknown = parser.parse_known_args(argv)
+
+    if not args.endpoint:
+        raise NoEndpointProvided(argv)
+    if args.endpoint in shortcuts and not Endpoint.is_endpoint(args.endpoint):
+        raise NoEndpointProvided(argv)
 
-    args, unknown = parser.parse_known_args(argv[:endpoint_index])
```
Comment (jayvdb) [inline jgo/jgo.py:726]: I've found a regression caused by this.  If the program args contains hyphenated args, e.g. `["com.pinterest:ktlint", "-F", "/c/path/to/file.kt"]` , the hyphenated arg becomes a JVM arg.

Comment (jayvdb) [reply jgo/jgo.py:726]: A workaround for this is for people to use `--` , i.e. `["com.pinterest:ktlint", "--", "-F", "/c/path/to/file.kt"]`
Requiring that workaround would mean this is a breaking change.
Looking for other solutions ...

Comment (jayvdb) [reply jgo/jgo.py:726]: Using `--` isnt enough.
`["--add-opens", "java.base/java.lang=ALL-UNNAMED", "com.pinterest:ktlint", "--", "-F", "/c/path/to/file.kt"]` ends up with jvm arg `["--add-opens"]` and endpoint `java.base/java.lang=ALL-UNNAMED`. :-(

**Comment by ctrueden** [comment]
@jayvdb Do you think the merge of #74 is sufficient to address your needs here? I.e. can this PR be closed? Or are there still benefits to updating and merging it?

**Comment by jayvdb** [comment]
I think there are benefits in where I was going with this, but I want to first get tests in place for the current behaviour, before attempting to take this approach further.  I'll get a WIP up shortly.

**Comment by jayvdb** [comment]
@ctrueden, I am pretty confident that I can get this working correctly, resulting in no more attempt to parse the program_args and all the problems that can cause.  OTOH, complexity goes through the roof, and we'd be using 'private' parts of the `ArgumentParser` class (`_match_arguments_partial` and `_parse_optional`), and making `parse_args` stateful (with those new members of the `ArgumentParser`) which is contrary to its intent.

Should I keep going?

**Comment by ctrueden** [comment]
With the redesigned jgo, now on the main branch, we use click (actually rich-click!) to parse arguments. I'm super happy with it, but feedback and testing from others would be very welcome!
